clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -711,13 +711,11 @@ private int partition(ProducerRecord<K, V> record, byte[] serializedKey , byte[]
         Integer partition = record.partition();
         if (partition != null) {
             List<PartitionInfo> partitions = cluster.partitionsForTopic(record.topic());
-            int numPartitions = partitions.size();
+            int numPartitions = partitions.size() - 1;
```

Comment:
This change made the numPartitions variable name inconsistent with semantic meaning. How about changing line 716 to `if (partition < 0 || partition > numPartitions - 1)` instead?